src/poetry/publishing/uploader.py

```diff
@@ -50,6 +50,44 @@ def __init__(self, error: ConnectionError | HTTPError | str) -> None:
         super().__init__(message)
 
 
+def _in_any(needle: str, haystacks: list[str]) -> bool:
+    return any(needle in haystack for haystack in haystacks)
+
+
+def _ignore_existing(response: requests.Response, skip_existing: bool) -> bool:
+    """Try to guess if a failed upload is safe to ignore
+
+    :param response:
+        The response from attempting to upload ``package`` to a repository.
+    :param skip_existing:
+        If ``True``, use the status and content of ``response`` to determine if the
+        package already exists on the repository. If so, then a failed upload is safe
+        to ignore.
+    :return:
+        ``True`` if a failed upload can be safely ignored, otherwise ``False``.
+    """
+    # based on https://github.com/pypa/twine/blob/a6dd69c79f7b5abfb79022092a5d3776a499e31b/twine/commands/upload.py#L32  # noqa: E501
+    if not skip_existing:
+        return False
+
+    status = response.status_code
+    reason = response.reason.lower()
+    text = response.text.lower()
+
+    return (
+        # pypiserver (https://pypi.org/project/pypiserver)
+        status == 409
+        # PyPI / TestPyPI / GCP Artifact Registry
+        or (status == 400 and _in_any("already exist", [reason, text]))
+        # Nexus Repository OSS (https://www.sonatype.com/nexus-repository-oss)
+        or (status == 400 and _in_any("updating asset", [reason, text]))
+        # Artifactory (https://jfrog.com/artifactory/)
+        or (status == 403 and "overwrite artifact" in text)
+        # Gitlab Enterprise Edition (https://about.gitlab.com)
+        or (status == 400 and "already been taken" in text)
```

Comment:
thoughts on just concatenating `reason` and `text` and going back to `in`? (a bit hacky and not strictly correct but perhaps good enough and simpler code)

Reply:
I've thought about it too and was a little torn. I like the simplicity but hate the hackiness. I think the "not strictly correct" argument can be neglected because such a false positive has to look really weird. Since I'm not the only one coming up with this idea, it may be fine here.